### PR TITLE
masking 

### PR DESCRIPTION
**Pull request by abuts**
These small changes fix masking in memory and tests for masking  and add `force_filebacked` option to sqw constructor, which is automaticlly appears if you specify `filebacked = true/false`. 
In addition it provides the test which demonstrates the issue #1081

### CODE REVIEW

File: _test/test_sqw_pixels/test_mask.m [hunk 1]
```diff
@@ -270,46 +222,139 @@ function test_mask_pixels_removes_pixels_given_in_mask_array(obj)
...
             new_sqw = mask_random_fraction_pixels(sqw_obj, frac_to_keep);
 
             expected_num_pix = round(frac_to_keep*sqw_obj.pix.num_pixels);
             assertEqual(new_sqw.pix.num_pixels, expected_num_pix);
+
+            test_mask.check_filebacked_signal_averages(new_sqw);
+
+        end
+        function test_mask_random_retains_correct_number_of_pix_on_file(obj)
+            skipTest('Re #1081 The test issustrates broken filebacked masking. Should be enabled as part of fixing the ticket')
```
Comment (cmarooney-stfc) [inline _test/test_sqw_pixels/test_mask.m:298]: add to the PR description that the PR does NOT fix #1081 so people are not confused by this

File: horace_core/sqw/@sqw/mask_pixels.m [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
+%   >> wout = mask_pixels (win, keep_obj)     % Mask array
 %   >> wout = mask_pixels (win, wmask)          % Mask according to pixel array
 %                                               % contents
 %
 % Input:
 % ------
 %   win                 Input sqw object
 %
-%   mask_array          Array of 1 or 0 (or true or false) that indicate
+%   keep_obj          Array of 1 or 0 (or true or false) that indicate
```
Comment (cmarooney-stfc) [inline horace_core/sqw/@sqw/mask_pixels.m:12]: maybe call it keep_pixels? Object sounds (a) vague (b) not a pixel

File: horace_core/sqw/@sqw/private/init_sqw_from_file_.m [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    else
+        error('HORACE:sqw:invalid_argument',...
+            ['Field "file" of init_from_file_ routine can ' ...
+            'contain only filename or initialized instance of loader class.\n' ...
+            ' Provided: %s'],disp2str(in_struc.file))
+    end
+else
+    error('HORACE:sqw:invalid_argument',...
+        ['init_from_file_ accepts only structure with field "file"' ...
+        ' containign filename or initialized data loader class.\n Provided: %s'], ...
```
Comment (cmarooney-stfc) [inline horace_core/sqw/@sqw/private/init_sqw_from_file_.m:19]:
```suggestion
        ' containing filename or initialized data loader class.\n Provided: %s'], ...
```

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

File: _test/test_sqw_pixels/test_mask.m (hunk 1 above)
Comment (abuts) [reply _test/test_sqw_pixels/test_mask.m:298]: hopefylly done

File: _test/test_sqw_pixels/test_mask.m [hunk 4]
```diff
@@ -270,46 +222,139 @@ function test_mask_pixels_removes_pixels_given_in_mask_array(obj)
...
-            file_info = dir(file_path);
-            %new_pg_size = file_info.bytes/6;
+        function [paged_sqw, masked_sqw,clOb] = get_paged_sqw(file_path, mask_array)
 
             paged_sqw = sqw(file_path,'file_backed', true);
+            new_pg_size = floor(paged_sqw.pix.num_pixels/6);
+            hc = hor_config;
+            mcs = hc.mem_chunk_size;
+            clOb = onCleanup(@()set(hc,'mem_chunk_size',mcs));
+            hc.mem_chunk_size = new_pg_size;
```
Comment (oerc0122) [inline _test/test_sqw_pixels/test_mask.m:353]: Use `set_temporary_config_options`

File: herbert_core/utilities/classes/@unique_references_container/unique_references_container.m [hunk 5]
```diff
@@ -413,13 +413,13 @@
         end
 
         %
-        function [self, nuix] = add_single_(self,obj)
+        function [self, nuix] = add_single_(self,inobj)
```
Comment (oerc0122) [inline herbert_core/utilities/classes/@unique_references_container/unique_references_container.m:416]: What is the purpose of renaming this?

File: horace_core/sqw/@sqw/mask.m [hunk 6]
```diff
@@ -32,12 +32,12 @@
...
-
 % Section the pix array, if non empty, and update pix_range
 if has_pixels(win)
     if wout.pix.is_filebacked
         wout = wout.get_new_handle();
     end
     wout.pix = wout.pix.mask(mask_array, win.data.npix);
 end
+% mask appropriate data
+wout.data = mask(wout.data,mask_array);
```
Comment (oerc0122) [inline horace_core/sqw/@sqw/mask.m:43]: Why does this need to move?

File: horace_core/sqw/@sqw/mask_pixels.m [hunk 7]
```diff
@@ -1,24 +1,24 @@
-function wout = mask_pixels (win, mask_array)
+function wout = mask_pixels (win, keep_obj)
```
Comment (oerc0122) [inline horace_core/sqw/@sqw/mask_pixels.m:1]: `keep_obj` is a bad name. Maybe `pix_to_keep`, honestly `mask_array` was probably best

File: horace_core/sqw/@sqw/private/init_sqw_from_file_.m [hunk 8]
```diff
@@ -0,0 +1,35 @@
+function obj = init_sqw_from_file_(obj, in_struc)
```
Comment (oerc0122) [inline horace_core/sqw/@sqw/private/init_sqw_from_file_.m:1]: Is this related to masking?

File: horace_core/sqw/PixelData/@PixelDataFileBacked/mask.m [hunk 9]
```diff
@@ -52,29 +52,40 @@
...
+page = 1;
+npix = obj.num_pixels;
+for i = 1:mem_chunk_size:npix
+    obj.page_num = page;
+    block_size = min(mem_chunk_size,npix-i+1);
+    if islogical(keep_array)
+        page_keep = keep_array(i:i+block_size-1);
+    else
+        selection = keep_array >=i & keep_array <i+block_size;
+        page_keep = keep_array(selection);
```
Comment (oerc0122) [inline horace_core/sqw/PixelData/@PixelDataFileBacked/mask.m:70]: This is not a `logical array 0 or 1` then. The function is specified to work with such.

Comment (oerc0122) [reply horace_core/sqw/PixelData/@PixelDataFileBacked/mask.m:70]: This also would mean that filebacked are masked differently to memory backed

**Review by oerc0122 — CHANGES_REQUESTED** [review]
File backed and memory masking are now different operations. 

Mask no longer agrees with documentation.

File: horace_core/sqw/@sqw/mask_pixels.m (hunk 2 above)
Comment (abuts) [reply horace_core/sqw/@sqw/mask_pixels.m:12]: according to the description, it may be `PixelData` object too. This option would certainly not work (for ticket), but naming needs to reflect this. Probably rename it to `keep_info`

File: herbert_core/utilities/classes/@unique_references_container/unique_references_container.m (hunk 5 above)
Comment (abuts) [reply herbert_core/utilities/classes/@unique_references_container/unique_references_container.m:416]: I am naturally perceive `obj` as `this` (i.e. the instance of object, the method called from) It is not the case here, so the renaming done to avoid the confusion

File: horace_core/sqw/@sqw/mask.m (hunk 6 above)
Comment (abuts) [reply horace_core/sqw/@sqw/mask.m:43]: This was the serious bug and this is the fixture. You mask pixel on the basis of non-masked npix as they define cells to mask. After you have dropped the pixels, you need to nullify the npix

File: horace_core/sqw/@sqw/mask_pixels.m (hunk 7 above)
Comment (abuts) [reply horace_core/sqw/@sqw/mask_pixels.m:1]: modified according to review above

Comment (abuts) [reply horace_core/sqw/@sqw/mask_pixels.m:1]: `mask_array` assumes that you remove data where the mask is true. The code does opposite -- this is why the removing done.

File: horace_core/sqw/@sqw/private/init_sqw_from_file_.m (hunk 8 above)
Comment (abuts) [reply horace_core/sqw/@sqw/private/init_sqw_from_file_.m:1]: it is related to the `test_mask` as the test_mask behaves differently depending on the size of the mem_chunk, testing either filebacked or memory backed pixels as function of memory chunk. The change force test to validate specific part of the code

File: horace_core/sqw/PixelData/@PixelDataFileBacked/mask.m (hunk 9 above)
Comment (abuts) [reply horace_core/sqw/PixelData/@PixelDataFileBacked/mask.m:70]: The description have been modified

Comment (abuts) [reply horace_core/sqw/PixelData/@PixelDataFileBacked/mask.m:70]: After some thinking, I've decided to prohibit non-logical indexes. There no direct request to implement them and I do not have time to write additional tests to fix the issue. 
Fixed

File: _test/test_sqw_pixels/test_mask.m [hunk 10]
```diff
@@ -270,46 +222,139 @@ function test_mask_pixels_removes_pixels_given_in_mask_array(obj)
...
             new_sqw = mask_random_fraction_pixels(sqw_obj, frac_to_keep);
 
             expected_num_pix = round(frac_to_keep*sqw_obj.pix.num_pixels);
             assertEqual(new_sqw.pix.num_pixels, expected_num_pix);
+
+            test_mask.check_filebacked_signal_averages(new_sqw);
+
+        end
+        function test_mask_random_retains_correct_number_of_pix_on_file(obj)
+            skipTest('Re #1081 The test illustrates broken filebacked masking. Should be enabled as part of fixing the ticket')
```
Comment (oerc0122) [inline _test/test_sqw_pixels/test_mask.m:298]: Why is this new test disabled when this is supposed to fix this issue?

File: horace_core/sqw/@sqw/mask_pixels.m [hunk 11]
```diff
@@ -1,24 +1,24 @@
-function wout = mask_pixels (win, mask_array)
+function wout = mask_pixels (win, keep_info)
```
Comment (oerc0122) [inline horace_core/sqw/@sqw/mask_pixels.m:1]: You now have a confusing mix of
`keep_array`
`keep_info`
`keep_obj`
scattered through the code. If I had to, I would pick `keep_array` but still prefer `mask_array`

File: horace_core/sqw/@sqw/mask_random_pixels.m [hunk 12]
```diff
@@ -25,7 +25,8 @@
 % Modifications: R. A. Ewings
 
 if ~has_pixels(win)
-    error('Can mask pixels only in an sqw-type object')
+    error('HORACE:sqw:invalid_argument', ...
+        'Can mask pixels only in an sqw-type object')
```
Comment (oerc0122) [inline horace_core/sqw/@sqw/mask_random_pixels.m:29]: This is kind of by definition an `sqw-type` object. This error message is now confusing. REfer to having zero pixels instead

File: horace_core/sqw/file_io/@horace_binfile_interface/private/parse_get_sqw_args_.m [hunk 13]
```diff
@@ -22,7 +26,8 @@
     'nopix', ...
     'legacy', ...
     'sqw_struc',...
-    'file_backed'
+    'file_backed',...
+    'force_pix_location'...
```
Comment (oerc0122) [inline horace_core/sqw/file_io/@horace_binfile_interface/private/parse_get_sqw_args_.m:30]: `force_pix_location` doesn't give any indication about what this does to me.

**Review by oerc0122 — CHANGES_REQUESTED** [review]

File: horace_core/sqw/@sqw/mask_pixels.m (hunk 11 above)
Comment (abuts) [reply horace_core/sqw/@sqw/mask_pixels.m:1]: `mask_array` assumes that `true` masks correspondent value while `keep_array` -- otherwise -- as it is in the code. 
`keep_array` is used where it is array. Here it may be `PixelData` object (according to description) Need to write test for that and may be fix, but this is another ticket.

File: _test/test_sqw_pixels/test_mask.m (hunk 10 above)
Comment (abuts) [reply _test/test_sqw_pixels/test_mask.m:298]: This test do not fix the issue. It demonstrates it. Can it be interpreted differently?

File: horace_core/sqw/file_io/@horace_binfile_interface/private/parse_get_sqw_args_.m (hunk 13 above)
Comment (abuts) [reply horace_core/sqw/file_io/@horace_binfile_interface/private/parse_get_sqw_args_.m:30]: well, do other names indicate?

Comment (oerc0122) [reply horace_core/sqw/file_io/@horace_binfile_interface/private/parse_get_sqw_args_.m:30]: `nopix` presumably gives me no pixels
`legacy` is presumably some legacy compatibility feature
`sqw_struc` presumably returns a structure rather than an sqw object
`file_backed` makes the pixels file backed.

Comment (abuts) [reply horace_core/sqw/file_io/@horace_binfile_interface/private/parse_get_sqw_args_.m:30]: I have added the detailed description of the purpose of this key. Its difficult to make short abbreviation on what its intedned to do.
